File: Tree/BinarySearchTreeIterator/BinarySearchTreeIterator.cc
/*
class TreeNode {
	int val;
	TreeNode left, right;
	TreeNode(int v) {
		this.left = NULL;
		this.right = NULL;
		this.val = v
	}
}
*/
class BSTIterator {
    private:
        Stack<TreeNode*> stack;
        TreeNode *curr = NULL;
    
    public:
        BSTIterator(TreeNode *root) {
            curr = root;
        } 
        
        bool hasNext() {
            return (curr && !stack.empty());
        }   
        
        TreeNode *next() {
            while(curr) {
                stack.push(curr);
                curr = curr->left
            }
            
            curr = stack.top();
            stack.pop();
            TreeNode* node = curr;
            curr = curr->right;
            
            return node;
            
        }
};

File: Tree/BinarySearchTreeIterator/BinarySearchTreeIterator.py
"""
class TreeNode:
    def __init__(self, v):
        self.val = v
        self.left, self.right = None, None 
"""
class BSTIterator:
    def __init__(self, s, root):
        self.stack = s
        self.curr = root
   
    def hasNext(self):
        return (curr && stack)
        
    def next(self) 
        while curr:
            stack.push(curr)
            curr = curr.left
            
        curr = stack.top()
        stack.pop()
        node = curr
        curr = curr.right
        
        return node
                
    
        
       
